src/openms/source/ANALYSIS/OPENSWATH/MRMDecoy.cpp

```diff
@@ -258,61 +265,70 @@ namespace OpenMS
+      // Block tryptic residues and N-/C-terminus from shuffling
+      MRMDecoy::IndexType idx = MRMDecoy::findFixedResidues(peptide.sequence, keepN, keepC, const_pattern);
 
-    for (Size j = 0; j < peptide.mods.size(); j++)
-    {
-      for (Size k = 0; k < peptide_index.size(); k++)
+      std::vector<Size> peptide_index;
+      for (Size i = 0; i < peptide.sequence.size(); i++)
       {
-        if (boost::numeric_cast<int>(peptide_index[k])  == peptide.mods[j].location)
-        {
-          peptide.mods[j].location = boost::numeric_cast<int>(k);
-          break;
-        }
+        peptide_index.push_back(i);
       }
-    }
 
-    return peptide;
-  }
+      // we erase the indices where K/P/R are (from the back / in reverse order
+      // to not delete indices we access later)
+      for (IndexType::reverse_iterator it = idx.rbegin(); it != idx.rend(); ++it)
+      {
+        peptide_index.erase(peptide_index.begin() + *it);
+      }
 
-  OpenMS::TargetedExperiment::Peptide MRMDecoy::reversePeptide(
-    OpenMS::TargetedExperiment::Peptide peptide)
-  {
-    OpenMS::TargetedExperiment::Peptide peptideorig = peptide;
-    std::vector<Size> peptide_index;
-    for (Size i = 0; i < peptide.sequence.size(); i++)
-    {
-      peptide_index.push_back(i);
-    }
+      // reverse the peptide index
+      std::reverse(peptide_index.begin(), peptide_index.end());
 
-    peptide.sequence = peptide.sequence.reverse();
-    std::reverse(peptide_index.begin(), peptide_index.end());
+      // re-insert the missing K/P/R at the appropriate places
+      for (IndexType::iterator it = idx.begin(); it != idx.end(); ++it)
+      {
+        peptide_index.insert(peptide_index.begin() + *it, *it);
+      }
 
-    for (Size j = 0; j < peptide.mods.size(); j++)
-    {
-      for (Size k = 0; k < peptide_index.size(); k++)
+      // use the reversed index to create the new peptide sequence and
+      // then to place the modifications at their appropriate places (make sure
+      // that the modifications are placed with their initial amino acids).
+      for (Size i = 0; i < peptide_index.size(); i++)
+      {
+        reversed.sequence[i] = peptide.sequence[peptide_index[i]];
+      }
+      for (Size j = 0; j < reversed.mods.size(); j++)
       {
-        if (boost::numeric_cast<int>(peptide_index[k]) == peptide.mods[j].location)
+        for (Size k = 0; k < peptide_index.size(); k++)
         {
-          peptide.mods[j].location = boost::numeric_cast<int>(k);
-          break;
+          // C and N terminal mods are implicitly not reversed because they live at positions -1 and sequence.size()
+          if (boost::numeric_cast<int>(peptide_index[k]) == reversed.mods[j].location)
+          {
+            reversed.mods[j].location = boost::numeric_cast<int>(k);
+            break;
+          }
         }
       }
+      return reversed;
     }
+  }
+
+  OpenMS::TargetedExperiment::Peptide MRMDecoy::pseudoreversePeptide(
```

Comment:
is `pseudo` a common term in this context?
If not, maybe there is something more descriptive? (reversePeptideConstCTerm()?)

Reply:
we have been using this in the targeted proteomics world quite a bit, so I opt to leave it.